src/connection.rs

```diff
@@ -92,438 +92,322 @@ impl<T: AsyncRead + AsyncWrite> Connection<T> {
-                if let OpCode::Close = frame.header().opcode() {
-                    self.state = Some(State::Closed);
-                    return Ok(Async::Ready(()))
-                }
-            Async::Ready(None) => self.state = Some(State::Closed),
-            Async::NotReady => self.state = Some(State::AwaitClose)
+    /// Write final header and payload data to socket.
+    ///
+    /// The data will be masked if necessary.
+    /// No extensions will be applied to header and payload data.
+    async fn write(&mut self, header: &mut Header, data: &mut [u8]) -> Result<(), Error> {
+        if self.mode.is_client() {
+            header.set_masked(true);
+            header.set_mask(rand::random());
+            self.codec.apply_mask(&header, data)
         }
-        Ok(Async::NotReady)
+        header.set_payload_len(data.len());
+        trace!("send: {}", header);
+        if self.wbuffer.len() > BLOCK_SIZE {
+            self.socket.write_all(&self.wbuffer).await?;
+            trace!("wrote {} bytes", self.wbuffer.len());
+            self.wbuffer.clear()
+        }
+        let header_bytes = self.codec.encode_header(&header);
+        self.wbuffer.extend_from_slice(header_bytes);
+        self.wbuffer.extend_from_slice(data);
+        Ok(())
     }
-}
 
-#[derive(Debug)]
-enum State {
-    /// Default state.
-    /// Possible transitions: `Open`, `AnswerPing`, `AnswerClose`, `Closed`.
-    Open(Option<base::Data>),
-
-    /// Send a PONG frame as answer to a PING we have received.
-    /// Possible transitions: `AnswerPing`, `Open`.
-    AnswerPing(Frame, Option<base::Data>),
-
-    /// Flush some frame we started sending.
-    /// Possible transitions: `Flush`, `Open`.
-    Flush(Option<base::Data>),
-
-    /// We want to send a close frame.
-    /// Possible transitions: `SendClose`, `FlushClose`.
-    SendClose(Frame),
-
-    /// We have sent a close frame and need to flush it.
-    /// Possible transitions: `FlushClose`, `AwaitClose`.
-    FlushClose,
-
-    /// We have sent a close frame and awaiting a close response.
-    /// Possible transitions: `AwaitClose`, `Closed`.
-    AwaitClose,
-
-    /// We have received a close frame and want to send a close response.
-    /// Possible transitions: `AnswerClose`, `Closing`.
-    AnswerClose(Frame),
-
-    /// We have begun sending a close answer frame and need to flush it.
-    /// Possible transitions: `Closing`, `Closed`.
-    Closing,
-
-    /// We are closed (terminal state).
-    /// Possible transitions: none.
-    Closed
-}
+    /// Receive the next websocket message.
+    ///
+    /// Fragmented messages will be concatenated and returned as one block.
+    /// The `bool` indicates if the data is textual (`true`) or binary
+    /// (`false`). If `Connection::validate_utf8` is `true` textual data
+    /// is checked for well-formed UTF-8 encoding before returned.
+    pub async fn receive(&mut self) -> Result<(BytesMut, bool), Error> {
```

Comment:
Is there a reason to have these methods instead of implementing `Sink` and `Stream`? The latter would allow using all sorts of stuff from `SinkExt`, `TrySinkExt` and so on. Also could be handy for using stuff like `select!`

Reply:
You can't use async/await when implementing `Sink`/`Stream` (or `AsyncRead`/`AsyncWrite` for that regard), which makes writing the implementation way more complex.

Reply:
Lack of async/await support is probably the biggest reason.

Reply:
GitHub should get better at updating comments while one is viewing the page. Seeing updates only after a manual refresh is so 90s.

Reply:
> You can't use async/await when implementing `Sink`/`Stream` (or `AsyncRead`/`AsyncWrite` for that regard), which makes writing the implementation way more complex.

But also makes usage more simple? At least that's the way I see this tradeoff.

Reply:
I was thinking more about how to integrate this with `Stream` and `Sink` and came up with this: https://github.com/twittner/soketto/compare/develop...twittner:into_stream_and_sink. Would it be acceptable to be able to turn a `Connection` into a pair of `Stream` and `Sink`?